templates/flutter/lib/src/realtime_mixin.dart.twig

```diff
@@ -0,0 +1,120 @@
+import 'dart:async';
+import 'dart:convert';
+import 'package:web_socket_channel/web_socket_channel.dart';
+import 'package:web_socket_channel/status.dart';
+import 'exception.dart';
+import 'realtime_subscription.dart';
+import 'client.dart';
+import 'realtime_message.dart';
+
+typedef Future<WebSocketChannel> WebSocketFactory(Uri uri);
+
+mixin RealtimeMixin {
+  late Client client;
+  Map<String, List<StreamController<RealtimeMessage>>> _channels = {};
+  WebSocketChannel? _websok;
+  String? _lastUrl;
+  late WebSocketFactory getWebSocket;
+
+  _closeConnection() {
+    _websok?.sink.close(normalClosure);
+  }
+
+  _createSocket() async {
+    final uri = _prepareUri();
+    if (_websok == null) {
+      await getWebSocket(uri);
+    }
+    if (_lastUrl == uri.toString() && _websok?.closeCode == null) {
+      return;
+    }
+    _closeConnection();
+    _lastUrl = uri.toString();
+    print('subscription: $_lastUrl');
+    _websok = await getWebSocket(uri);
+
+    try {
+      _websok?.stream.listen((event) {
+        final data = jsonDecode(event);
+        handleError(data);
+        if(data.containsKey('event')) {
+          final message = RealtimeMessage.fromMap(data);
+          message.channels.forEach((channel) {
+            if (this._channels[channel] != null) {
+              this._channels[channel]!.forEach((stream) {
+                stream.sink.add(message);
+              });
+            }
+          });
+        }
+      });
+    } catch (e) {
+      if (e is {{spec.title | caseUcfirst}}Exception) {
+        rethrow;
+      }
+      if (e is WebSocketChannelException) {
+        throw {{spec.title | caseUcfirst}}Exception(e.message);
+      }
+      throw {{spec.title | caseUcfirst}}Exception(e.toString());
+    }
+  }
+
+  Uri _prepareUri() {
+    if (client.endPointRealtime == null) {
+      throw {{spec.title | caseUcfirst}}Exception(
+          "Please set endPointRealtime to connect to realtime server");
+    }
+    var uri = Uri.parse(client.endPointRealtime!);
+    return Uri(
+      host: uri.host,
+      scheme: uri.scheme,
+      port: uri.port,
+      queryParameters: {
+        "project": client.config['project'],
+        "channels[]": _channels.keys.toList(),
+      },
+      path: uri.path + "/realtime",
+    );
+  }
+
+  RealtimeSubscription subscribeTo(List<String> channels) {
+    StreamController<RealtimeMessage> controller = StreamController.broadcast();
+    channels.forEach((channel) {
+      if (!this._channels.containsKey(channel)) {
+        this._channels[channel] = [];
+      }
+      this._channels[channel]!.add(controller);
+    });
+    Future.delayed(Duration.zero, () => _createSocket());
+    RealtimeSubscription subscription = RealtimeSubscription(
+        stream: controller.stream,
+        close: () {
+          controller.close();
```

Comment:
Is it possible to have the same method signature as Web and Kotlin?

Meaning `subscribe(channel, callback)`

Reply:
But, stream is the proper way to listen to live changes in Flutter/Dart, I don't think using callback would give proper experience.